Experiment automation/(LUMICKS) Reference kit automation script/Catch_beads_DNA_record_Kymo_for_LacI_version_1.py

```diff
@@ -0,0 +1,318 @@
+    ie make sure the correct location is being scanned when making a kymograph
+    4) Make sure the traps are not too close together for catching beads
+
+Workflow of the script
+    1) Catch beads
+    2) Catch a DNA tether
+    3) Check if it is only a single tether, if not, try to break tethers until there is 1 left.
+    4) Move using the microstage to 'J1' at 60% speed
+    5) Move using the microstage to protein channel at 60% speed
+    6) Set the force to 10 pN
+    7) Start a kymograph.
+    8) Stop the kymograph.
+    9) Restart step 1 to 8 n times, where n is the number of kymographs you want to make.
+"""
+from datetime import datetime
+import bluelake as bl
+import time
+import os
+import numpy as np
+
+
+name_bead_channel = "beads"  # This has to match the correct channel name in the UI
+name_dna_channel = "DNA"  # This has to match the correct channel name in the UI
+name_protein_channel = "Ch4"
+name_junction = "J1"
+name_buffer_channel = "buffer"
+time_to_wait_for_flow = 5.0
+dna_length = 0.34*37.8  # Length of DNA in micrometers
+
+distance = bl.timeline["Distance"]["Distance 1"]
+force = bl.timeline["Force LF"]["Trap 2"]
+match_score1 = bl.timeline["Tracking Match Score"]["Bead 1"]
+match_score2 = bl.timeline["Tracking Match Score"]["Bead 2"]
+
+
+def throw_if_beads_lost(match_threshold):
+    """Raise an exception if we lose the beads."""
+    if match_score1.latest_value < match_threshold or match_score2.latest_value < match_threshold:
+        raise RuntimeError("Lost beads")
+
+
+def set_pressure(target):
+    """Increase pressure until we are above a certain target level"""
+    while bl.fluidics.pressure < target:
+        bl.fluidics.increase_pressure()
+        bl.pause(1.0)
+
+
+def start_flow(pressure):
+    print("Starting Flow.")
+    set_pressure(pressure)
+    bl.fluidics.open(1, 2, 3, 6)
+
+
+def stop_flow():
+    print("Stopping flow.")
+    bl.fluidics.close(1, 2, 3, 6)
+
+
+def catch_beads(match_threshold, pressure=0.25):
+    """Starts the flow and attempts to catch beads. Toggles the shutters when match score is too low."""
+    start_flow(pressure)
+
+    print(f"Waiting {time_to_wait_for_flow} sec for flow to begin.")
+    bl.pause(time_to_wait_for_flow)
+
+    print("Moving to bead channel.")
+    bl.microstage.move_to(name_bead_channel)
+    bl.pause(1.0)
+    bl.mirror1.clear()
+    bl.mirror2.clear()
+
+    print("Trapping beads.")
+    start_time = time.time()
+    n = 0
+    while match_score1.latest_value < match_threshold or match_score2.latest_value < match_threshold:
+        """Drop beads that do not fulfill the template"""
+        if 0 < match_score1.latest_value < match_threshold:
+            bl.mirror1.clear()
+        if 0 < match_score2.latest_value < match_threshold:
+            bl.mirror2.clear()
+        if n > 50:
+            stop_flow()
+            raise Exception("Need more than 750 seconds to catch beads, stop script")

```

Comment:
Usually good habit to be as specific as possible with exceptions.
```suggestion
            raise RuntimeException("Needed more than 750 seconds to catch beads, stopping script")
```